content/assignments/minishell/checker/_test/run_test.sh

```diff
@@ -0,0 +1,244 @@
+	fi
+
+	if [ "$USE_VALGRIND" = "yes" ]; then
+		which valgrind &> /dev/null
+		if [ $? -eq 0 ]; then
+			exec_name="valgrind --leak-check=full \
+			                    --show-leak-kinds=all \
+					    --child-silent-after-fork=yes \
+					    --track-fds=yes \
+					    --log-file=$VALGRIND_LOG \
+					    $exec_name"
+		else
+			echo "valgrind is not installed."
+		fi
+	fi
+
+	# Generates random file as test file.
+	TEST_NAME=$(printf "test_%02d" "$test_index")
+	OUT_DIR="${TEST_NAME}_out"
+	REF_DIR="${TEST_NAME}_ref"
+	IN_FILE="${TEST_NAME}.in"
+	REF_FILE="${TEST_NAME}.ref"
+	OUT_FILE="${TEST_NAME}.out"
+
+	local input_basename="${INPUT_DIR}/${TEST_NAME}"
+	[ -f "${input_basename}.${OS_PLATFORM}.txt" ] && \
+		input_basename="${input_basename}.${OS_PLATFORM}" || \
+		input_basename="${input_basename}"
+
+	# Copy input files in the testing directory.
+	cp "${input_basename}.txt" "${MAIN_TEST_DIR}/${IN_FILE}" &> "$LOG_FILE"
+	# Copy additional files, if they exist (see test 04).
+	# cp ${input_basename}?*.txt ${MAIN_TEST_DIR} &> $LOG_FILE
+	# Copy any reference files, if they exist (see test 18).
+	cp "${REFS_DIR}/${TEST_NAME}.ref" "${MAIN_TEST_DIR}" &> "$LOG_FILE"
+
+	# Export mini-shell in path.
+	PATH="$PATH:$(pwd)"
+	cd "${MAIN_TEST_DIR}" &> "$LOG_FILE" || exit 1
+
+	[ -d "${OUT_DIR}" ] && rm -rf "${OUT_DIR}"
+	[ -d "${REF_DIR}" ] && rm -rf "${REF_DIR}"
+}
+
+# Cleanups a test.
+cleanup_test()
+{
+	if [ "$DO_CLEANUP" = "yes" ]; then
+		[ -d "${OUT_DIR}" ] && rm -rf "${OUT_DIR}"
+		[ -d "${REF_DIR}" ] && rm -rf "${REF_DIR}"
+		[ -e "${IN_BASH}" ] && rm "${IN_BASH}"
+		[ -e "${IN_MINI}" ] && rm "${IN_MINI}"
+	fi
+}
+
+# Initializes the whole testing environment.
+# Should be the first test called.
+init_world()
+{
+	[ -d "$MAIN_TEST_DIR" ] && rm -rf "$MAIN_TEST_DIR"
+	mkdir -p "${MAIN_TEST_DIR}"
+	print_header "Testing - mini-shell"
+}
+
+# Cleanups the whole testing environment.
+# Should be the last test called.
+cleanup_world()
+{
+	[ "$DO_CLEANUP" = "yes" ] && rm -rf "$MAIN_TEST_DIR"
+}
+
+# ----------------- Test Suite ----------------------------------------------- #
+
+# Specific test example.
+execute_cmd()
+{
+	EXEC=$1
+	INPUT=$2
+	OUTPUT=$3
+	[ -z "$OUTPUT" ] && OUTPUT=$LOG_FILE
+	mkdir -p "${OUT_DIR}" && cd "${OUT_DIR}" &> "$LOG_FILE" || exit 1
+	timeout "$TEST_TIMEOUT" $BUFFERING_WRAPPER $EXEC < "${INPUT}" &> "$OUTPUT"
+	# Ocasionally, in a virtualized environment, the diff that compares the
+	# target implementation with the reference output starts before the
+	# target implementation has finished writing files to disk. Handle this
+	# by mandating that all I/O buffers be flushed.
+	sync
+	cd - &> "$LOG_FILE" || exit 1
+}
+
+
+generate_input()
+{
+	# Generate the PROMPT for BASH.
+	sed 's/\(.*\)/printf \"> \"; \1/' "${IN_FILE}" > "${REF_FILE}"
+}
+
+# Checks the output of the commands.
+test_output()
+{
+	init_test
+
+	generate_input
+
+	# Commands to execute the test.
+	execute_cmd "$ref_name" "../${REF_FILE}" "${OUT_FILE}"
```

Comment:
It's hard to follow why the `../${REF_FILE}` is needed here, but not in `generate_input`. I assume it's because of the `cd` in `execute_cmd`, but I'm not sure. It may be worth keeping the absolute path to the files (e.g. using `local ref=$(readlink -e "${REF_FILE})` and passing that as the argument.

Reply:
Agree.
I will make another issue and solve it with another PR.